### ISSUE TITLE
.obsidian.vimrc not working

### ISSUE THREAD

**Issue by LAST7** [esm7/obsidian-vimrc-support#122]
Hi, thanks for developing this awesome plugin, but I seem to have an issue:

Nothing in my `.obsidian.vimrc` seems to work, including `imap jj <Esc>` and settings like `nmap j gj` mentioned in the README.md. 
But when I type `:imap jj <Esc>` in the command mode, it works. I might have done something wrong, but I can not figure it out myself since I know little about how these things work, so I need your help.

Some detailed information about my situation:

1. win 10
2. working with the latest version of both Obsidian(0.15.6) and vimrc-support plugin(0.7.1)
3. option  "Use legacy editor"  turned OFF
4. enabled the built-in Vim support before installing the plugin and never disabled
5. created file named .obsidian.vimrc in the root directory of the current vault
6. some of my notes contains characters that are not English
7. enabled the "Vim chord display" and "Vim mode display" in the plugin settings, but nothing seems to change
8. my Obsidian Developer Console information after start up:
```
Obsidian Developer Console
plugin:obsidian-outliner:2499 Loading obsidian-outliner
plugin:url-into-selection:690 loading url-into-selection
plugin:quickadd:16120 Loading QuickAdd
plugin:obsidian-cursor-location-plugin:88 loading Cursor Location plugin
plugin:obsidian-git:14409 loading Obsidian Git plugin
plugin:obsidian-better-command-palette:15 Loading plugin: Better Command Palette
```
I don't see the plugin loaded in this log, but I don't know why did this happened.

**Comment by esm7** [comment]
Thank you for the detailed report.
Apparently in the latest version I unintentionally removed the log when the plugin loads, so not seeing it doesn't mean it's not loaded.

1. One thing you haven't mentioned, so I just want to make sure, have you enabled the Vimrc plugin itself (from the Community plugins list)? I assume so because you mentioned a few configuration options, but just wanna make sure.
2. With the Obsidian Developer Console open, can you try the command `:obcommand` and see what happens?
3. Can you post a screenshot that shows your `.obsidian.vimrc` file in its location, and post its full content?

**Comment by LAST7** [comment]
Thanks for your respond!

1. the vimrc-support plugin is enabled for sure, I've checked it again.
2. with the Obsidian Developer Console open, command `:obcommand` will get an error: `not an editor command`, maybe its the core of my issue?
3. screenshots:
![image](https://user-images.githubusercontent.com/96737659/180162463-40b06e88-1871-4012-b1f6-0f1320da75ec.png)
![image](https://user-images.githubusercontent.com/96737659/180162534-53c51695-ad90-484c-8541-584120b2e7c2.png)

**Comment by esm7** [comment]
The fact you don't have `:obcommand` registered shows that the Vimrc plugin is indeed not loaded at all.
1. Can you maybe try to uninstall & reinstall the plugin?
2. Are you absolutely sure that "Use legacy editor" is OFF?
3. What is your default editing mode - live preview or source mode?

**Comment by LAST7** [comment]
1. I tried uninstall & reinstall the plugin, but unfortunately the same issue still exist.
2. 100% percent sure that "Use legacy editor" is OFF, I turned the software language into English and checked again.
3. my default editing mode is live preview, but the issue still exist when I switched it into source mode.

**Comment by sudoLife** [comment]
I had the same problem until I upgraded to 0.15.6-1. It works again now.

**Comment by kohane27** [comment]
I'm sad to report that I'm having the same issue. 

Config:

Linux flatpak version 0.15.6
vimrc 0.71

**Comment by LAST7** [comment]
@sudoLife 
This sounds weird, maybe it has something to do with the software itself.
By the way, could you please tell me how to download Obsidian version 0.15.6-1? I could only find version 0.15.6 and 0.15.7 (insider build).

**Comment by esm7** [comment]
Can any of these who experience this issue check if it works if the plugin is downgraded to version 0.6.3 (I think it has to be done manually, as Obsidian doesn't have a UI for this)?
And if it works in 0.6.3, attach a console printout of how it looks in 0.6.3?

**Comment by LAST7** [comment]
@esm7 
Sadly, the same issue remain. Command `:obcommand` still gets an error: `Not an editor command`.
the console printout (I turned on a function which will print out the time that a plugin use to start up):
```
Obsidian Developer Console
ObsidianOutlinerPlugin.ts:36 Loading obsidian-outliner
main.ts:17 loading url-into-selection
plugin:quickadd:16120 Loading QuickAdd
main.ts:44 loading Cursor Location plugin
plugin:obsidian-git:14409 loading Obsidian Git plugin
plugin:obsidian-better-command-palette:15 Loading plugin: Better Command Palette
app.js:1 Total plugin setup: 130ms
obsidian-full-calendar: 56ms
obsidian-git: 26ms
obsidian-outliner: 20ms
quickadd: 13ms
obsidian-cursor-location-plugin: 4ms
obsidian-better-command-palette: 4ms
url-into-selection: 3ms
obsidian-vimrc-support: 3ms
obsidian-reveal-active-file: 1ms
main.ts:88 Vimrc plugin: using CodeMirror 6 mode
```
I didn't see the last line in the console printout with the 0.7.1 version of vimrc-support. But the third line to the last, which is `obsidian-vimrc-support: 3ms` **did show up** with the 0.7.1 version of vimrc-support.
Another fact is that the plugin settings panel does show up when I click the setting button, but nothing would change even I turned on these functions: `Vim chord display`, `Vim mode display`.

![image](https://user-images.githubusercontent.com/96737659/180599381-0585594a-f93f-4a64-8ba3-5e513a26378e.png)

**Comment by clearf** [comment]
I am having what might be a similar issue. Some things still work in vi-mode, but there are some commands that don't work (the inability to go to the beginning or end of a wrapped line is what caught my attention.)

Here's my obsidian vimrc file: 

`nmap j gj
nmap k gk
nmap 0 g0
nmap $ g$


" Have these be smart lines
" exmap lineStart cmcommand goLineLeftSmart
" exmap lineEnd cmcommand goLineRight
" nmap 0 :lineStart
" nmap $ :lineEnd


" Surround text with [[ ]] to make a wikilink
" NOTE: must use 'map' and not 'nmap'
exmap wiki surround [[ ]]
map [[ :wiki
`
And there's something strange that happens when I enter the vi command mode (whatever pressing `:` is called). 

For reference, here's my normal obsidian window: 

![image](https://user-images.githubusercontent.com/1944486/180634609-c9ac27b7-37e2-4683-a4a9-f5bd9f79dfe2.png)


But when I enter `:g$`, out pops a crazy unformatted text window in the middle of obsidian: 

![image](https://user-images.githubusercontent.com/1944486/180634595-af3444e4-4c61-42b9-9182-7b852ad0045c.png)

Thanks,
Chris

**Comment by LAST7** [comment]
New situation.
I just installed Obsidian (v0.15.8) with flatpak on my kali linux system and the plugin (v0.7.1) works fine and the command `:obcommand` doesn't get any error, instead the `Uncaught error: obcommand requires exactly 1 parameter` showed up in the printout of the Console. Everything works just fine.
But on my win 10 system, Obsidian (v0.15.8) with the newest plugin (0.7.1) still doesn't read any configuration from `.obsidian.vimrc`, and command `:obcommand` still gets the `Not an editor command` error.

So my solution to the issue might be simply installing WSL, and run Obsidian on my WSL...
Anyway if the key reason that leads to my issue lays in the win 10 system, I think it might be difficult to fix this bug.
Thanks for your time and attention, you've developed a wonderful and very practical plugin.
Feel free to close this issue if you want.

**Comment by esm7** [comment]
I'm hoping to release a debug build for you guys to run and get more information, but this last comment made me think of something. One of the latest Obsidian versions (I think it was 0.15.7?) was supposed to pop up a message saying this version requires a new installer.
Did you see this message, and did you indeed re-download & reinstall Obsidian with this version?
If not, can you try this direction?
Requiring WSL for Vimrc to work can't be a general solution for everybody, naturally, and we must get to the bottom of this.

**Comment by LAST7** [comment]
Yes the new version of Obsidian (v0.15.8) does require a new installer. I downloaded it from the browser and re-downloaded & reinstalled Obsidian.
But the exact same issue still remain on my win 10 system.

Thanks again for your attention and time, I'd be happy to provide more information to find a better solution to this issue.

**Comment by prurph** [comment]
I've also noticed this on Windows, and further have seen a working installation "break". It seems like it might be correlated with changing `.obsidian.vimrc` while Obsidian is open(?).

What I've seen happen is:

- Plugin is working correctly
- I make changes to `.obsidian.vimrc` and restart Obsidian
- Sometimes, this causes the plugin stops working, regardless of disable/re-enable, uninstall/reinstall, or reverting the vimrc

As if the above isn't confusing enough, in this state removing the vimrc file completely _will_ cause the plugin to complain that there's no vimrc, but then putting it back doesn't cause mappings therein to work.

The only fix I've found is:

- Uninstall the plugin and close Obsidian
- Create a very simple `.obsidian.vimrc` with just one line, I suggest `set clipboard=unnamed` because it will cause the Plugin to emit a log line to the dev console if it works
- Reinstall and re-enable the plugin. Restart Obsidian and you should see "Vim is now set to yank to system clipboard." in console
- Close Obsidian, swap in the old `.obsidian.vimrc` that wasn't working, restart Obsidian and now it works?

It feels like a filesystem issue and I am using iCloud Drive to store my vault so I wonder if there's some issue where the file gets in a weird state and can't be accessed by Obsidian.

**Comment by Graicc** [comment]
> The only fix I've found is:
> 
> * Uninstall the plugin and close Obsidian
> * Create a very simple `.obsidian.vimrc` with just one line, I suggest `set clipboard=unnamed` because it will cause the Plugin to emit a log line to the dev console if it works
> * Reinstall and re-enable the plugin. Restart Obsidian and you should see "Vim is now set to yank to system clipboard." in console
> * Close Obsidian, swap in the old `.obsidian.vimrc` that wasn't working, restart Obsidian and now it works?
> 
> It feels like a filesystem issue and I am using iCloud Drive to store my vault so I wonder if there's some issue where the file gets in a weird state and can't be accessed by Obsidian.

I had the same issue with a vault on OneDrive, this fixed it for me. I created a new vault on my desktop and the same vimrc worked without any issues, so I also suspect it is a filesystem issue.

**Comment by kohane27** [comment]
TL;DR disable [shichongrui/obsidian-reveal-active-file](https://github.com/shichongrui/obsidian-reveal-active-file) and `obsidian-vimrc-support` should work.

I created a new vault with just `obsidian-vimrc-support` installed and it works perfectly. So I disabled all my plugins in my main vault and `obsidian-vimrc-support` worked perfectly too. So there's some conflict among my plugins. After disabling some plugins I discovered that the culprit is [shichongrui/obsidian-reveal-active-file](https://github.com/shichongrui/obsidian-reveal-active-file). Disabling it and `obsidian-vimrc-support` worked every time. I have no idea why. If you don't have `shichongrui/obsidian-reveal-active-file` installed, try disabling other potential conflicting plugins that's causing `obsidian-vimrc-support` to stop working.

Cheers!

**Comment by prurph** [comment]
I've also noticed that [https://github.com/tgrosinger/recent-files-obsidian](tgrosinger/recent-files-obsidian) breaks `obsidian-vimrc-support`, so it seems like many plugins that also have on file open bindings interfere with the functionality of this plugin.

**Comment by LAST7** [comment]
I follow the suggestion by @prurph and @kohane27 , uninstalled the `obsidian-reveal-active-file` plugin and then the `vimrc-support` plugin now works perfectly.
So maybe my situation is that both the plugin conflict and the unkown filesystem issue happened on my Obsidian, and they are now solved.
Thanks to everyone here who provided suggestions, and thanks to the developer of this plugin.

I'm leaving this issue open, close it if you want.

**Comment by sudoLife** [comment]
@LAST7 , sorry I'm late to the party, but I have my version of obsidian from `obsidian-appimage` in AUR (Arch User Repository).

I see that you resolved your issue, that's great!

**Comment by prurph** [comment]
I think there's some non-determinism here--in trying to troubleshoot I re-added both `recent-files-obsidian` and `obsidian-reveal-active-file` and my vimrc is loading correctly.

**Comment by prurph** [comment]
Could anyone check out the changes from [my PR](https://github.com/esm7/obsidian-vimrc-support/pull/126) to see if they fix the issues? It's non-deterministic on my system so while it's working for me, I'm not sure if it's actually fixed (or that I didn't break anything!)

Testing would require node/npm and be something like:

```shell
$ cd <your vault location>/.obsidian/plugins
$ rm -rf obsidian-vimrc-support
$ git clone git@github.com:prurph/obsidian-vimrc-support.git && git checkout patch-1
$ npm install
$ npm run build
```

Then reopen Obsidian.

**Comment by kohane27** [comment]
@prurph  I tested your patch and I'm sad to report that it's not working.

To test:

1. Install back [`obsidian-reveal-active-file`](https://github.com/shichongrui/obsidian-reveal-active-file)
2. Restart Obsidian
3. Confirmed `.obsidian.vimrc` not working
4. git clone https://github.com/prurph/obsidian-vimrc-support.git to plugins dir
5. cd obsidian-vimrc-support
6. git checkout patch-1
7. Restart Obsidian
8. `.obsidian.vimrc` not working

To reliably reproduce this bug, you may install [`obsidian-reveal-active-file`](https://github.com/shichongrui/obsidian-reveal-active-file), then open the File Tree panel on the left when starting up Obsidian.

**Comment by esm7** [comment]
Can one of those experiencing the issue test the attached main.js file (by putting it in the `obsidian-vimrc-support` folder under `.obsidian/plugin` and restarting Obsidian)?

**Comment by prurph** [comment]
@kohane27 thanks for checking, and sorry to hear it didn't fix the issue. What's weird is those same steps work for me.

Perhaps it is as @esm7 said that other plugins are not using registerEvent, or it may be this doesn't address the issue at all and is simply a red herring.

It might be worth trying the same change in the reveal plugin to see if there is any effect.

**Comment by esm7** [comment]
@prurph can you try the vimrc.zip I attached in the previous comment?
I had a chat with Licat about this earlier, he believes it has nothing to do with the type of event registration, but gave other pointers which I implemented in that attempted fix.

**Comment by kohane27** [comment]
> @prurph can you try the vimrc.zip I attached in the previous comment? I had a chat with Licat about this earlier, he believes it has nothing to do with the type of event registration, but gave other pointers which I implemented in that attempted fix.

@esm7 not @prurph But I just unzipped and tried your fix. It worked even after enabling `obsidian-reveal-active-file`!

**Comment by esm7** [comment]
Awesome, if the others confirm the fix too I'll release it as a proper version later today.

**Comment by prurph** [comment]
@esm7 your main.js works, thanks!

**Comment by LAST7** [comment]
@esm7 Gread, that `main.js` works.
Neither editting `.obsidian.vimrc` when Obsidian is opened or installing `obsidian-reveal-active-file` would stop `vimrc-support` from working.

**Comment by esm7** [comment]
Released 0.7.2 with this fix. Thanks everybody for the help in pinpointing the issue here.
